gripper_action_controller/include/gripper_action_controller/gripper_action_controller.h

```diff
@@ -0,0 +1,199 @@
+  };
+
+  GRIPPER_ACTION_CONTROLLER_PUBLIC GripperActionController();
+
+  GRIPPER_ACTION_CONTROLLER_PUBLIC
+  controller_interface::return_type
+  init(const std::string &controller_name) override;
+
+  /**
+   * @brief command_interface_configuration This controller requires the
+   * position command interfaces for the controlled joints
+   */
+  GRIPPER_ACTION_CONTROLLER_PUBLIC
+  controller_interface::InterfaceConfiguration
+  command_interface_configuration() const override;
+
+  /**
+   * @brief command_interface_configuration This controller requires the
+   * position and velocity state interfaces for the controlled joints
+   */
+  GRIPPER_ACTION_CONTROLLER_PUBLIC
+  controller_interface::InterfaceConfiguration
+  state_interface_configuration() const override;
+
+  GRIPPER_ACTION_CONTROLLER_PUBLIC
+  controller_interface::return_type update() override;
+
+  GRIPPER_ACTION_CONTROLLER_PUBLIC
+  rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+  on_configure(const rclcpp_lifecycle::State &previous_state) override;
+
+  GRIPPER_ACTION_CONTROLLER_PUBLIC
+  rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+  on_activate(const rclcpp_lifecycle::State &previous_state) override;
+
+  GRIPPER_ACTION_CONTROLLER_PUBLIC
+  rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+  on_deactivate(const rclcpp_lifecycle::State &previous_state) override;
+
+  GRIPPER_ACTION_CONTROLLER_PUBLIC
+  rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+  on_cleanup(const rclcpp_lifecycle::State &previous_state) override;
+
+  GRIPPER_ACTION_CONTROLLER_PUBLIC
+  rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+  on_error(const rclcpp_lifecycle::State &previous_state) override;
+
+  GRIPPER_ACTION_CONTROLLER_PUBLIC
+  rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+  on_shutdown(const rclcpp_lifecycle::State &previous_state) override;
+
+  realtime_tools::RealtimeBuffer<Commands> command_;
+  // pre-allocated memory that is re-used to set the realtime buffer
+  Commands command_struct_, command_struct_rt_;
+
+private:
+  using GripperCommandAction = control_msgs::action::GripperCommand;
+  using ActionServer = rclcpp_action::Server<GripperCommandAction>;
+  using ActionServerPtr = ActionServer::SharedPtr;
+  using GoalHandle = rclcpp_action::ServerGoalHandle<GripperCommandAction>;
+  using RealtimeGoalHandle = realtime_tools::RealtimeServerGoalHandle<
+      control_msgs::action::GripperCommand>;
+  using RealtimeGoalHandlePtr = std::shared_ptr<RealtimeGoalHandle>;
+
+  using HwIfaceAdapter = HardwareInterfaceAdapter<HardwareInterface>;
+
+  bool update_hold_position_;
+
+  bool verbose_ = false; ///< Hard coded verbose flag to help in debugging
+  std::string name_;     ///< Controller name.
+  hardware_interface::LoanedCommandInterface *joint_position_command_interface_;
```

Comment:
I would really appreciate if we didn't have to use raw pointers in this new age of modern C++.
Here's the alternative https://github.com/ros-controls/ros2_controllers/blob/master/diff_drive_controller/include/diff_drive_controller/diff_drive_controller.hpp#L95

Reply:
The `reference_wrapper` isn't a default constructible type, as you can see we're setting up the interfaces in `on_activate`, the only reason why I used raw pointers because they don't own the objects, what do you about using `std::optional<std::size_t>` as an index for the given interface rather than pointers .?

Reply:
anything sounds much better than a raw pointer, let's go with the optional for now!